### PR TITLE
[CI] Update CI to include a Windows build of CppInterOp + other general CI improvements

### PR DESCRIPTION
**Pull request by mcbarton**
PR to add CI build infrastructure + fixes https://github.com/compiler-research/CppInterOp/issues/182

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -334,6 +444,62 @@ jobs:
...
+          git clone --depth=1 -b cling-llvm${{ matrix.clang-runtime }} https://github.com/root-project/llvm-project.git
+        }
+        else
+        {
+          git clone --depth=1 -b release/${{ matrix.clang-runtime }}.x https://github.com/llvm/llvm-project.git
+        }
+        cd llvm-project
+        # Apply patches
+        echo "Apply clang${{ matrix.clang-runtime }}-*.patch patches:"
+        compgen -G "../patches/llvm/clang${{ matrix.clang-runtime }}-*.patch" > /dev/null && find ../patches/llvm/clang${{ matrix.clang-runtime }}-*.patch -exec basename {} \; -print -exec git apply {} \;
```
Comment (mcbarton) [inline .github/workflows/ci.yml:466]: @vgvassilev this PR is expected to currently fail to build on Windows at the moment, as I don't know what the windows equivalent of this linux/osx command is. Any ideas?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -191,6 +277,37 @@ jobs:
         echo "LLVM_HASH=$LLVM_HASH" >> $GITHUB_ENV
     - uses: nelonoel/branch-name@v1.0.1
 
+    - name: Save PR Info on Windows systems
```
Comment (mcbarton) [inline .github/workflows/ci.yml:280]: Need to determine how to get CLING_HASH AND LLVM_HASH on Windows. Have idea on how to solve this. Will test tomorrow.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (mcbarton) [reply .github/workflows/ci.yml]: This has been solved now

**Comment by mcbarton** [comment]
@alexander-penev @vgvassilev Currently I've got the Windows ci build upto the point where it needs to build CppInterOp , but it fails to find the llvm config file. I've tried a few different definitions of LLVM_DIR without success. Any ideas what the folder location should be on Windows?

A current workaround that needs to be fixed for the Windows build is that LLVM_HASH and CLING_HASH are currently hardcoded. After these 2 issues have been fixed then this project should have a working ci which shows that CppInterOp builds on Windows, assuming no other issues arise once LLVM_DIR issue is resolved.

**Comment by mcbarton** [comment]
> @alexander-penev @vgvassilev Currently I've got the Windows ci build upto the point where it needs to build CppInterOp , but it fails to find the llvm config file. I've tried a few different definitions of LLVM_DIR without success. Any ideas what the folder location should be on Windows?
> 
> A separate issue which also needs to be fixed is that currently for the Windows ci build LLVM_HASH and CLING_HASH are currently hardcoded (I have an idea for a fix, but have yet to test). After these 2 issues have been fixed then this project should have a working ci which shows that CppInterOp builds on Windows, assuming no other issues arise once LLVM_DIR issue is resolved.

Update I have now solved the LLVM_DIR and Clang_DIR issue. Now I  need to solve the following issue with the cmake configure stage of building CppInterOp
```
-- Looking for os_signpost_interval_begin
CMake Error at D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2 (set):
  Syntax error in cmake code at

    D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2

  when parsing string

    D:/a/CppInterOp/CppInterOp/cmake;D:/a/CppInterOp/CppInterOp/cmake/modules;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:\a\CppInterOp\CppInterOp\llvm-project\build\lib\cmake\llvm

  Invalid character escape '\a'.
```

**Comment by vgvassilev** [comment]
> > @alexander-penev @vgvassilev Currently I've got the Windows ci build upto the point where it needs to build CppInterOp , but it fails to find the llvm config file. I've tried a few different definitions of LLVM_DIR without success. Any ideas what the folder location should be on Windows?
> > A separate issue which also needs to be fixed is that currently for the Windows ci build LLVM_HASH and CLING_HASH are currently hardcoded (I have an idea for a fix, but have yet to test). After these 2 issues have been fixed then this project should have a working ci which shows that CppInterOp builds on Windows, assuming no other issues arise once LLVM_DIR issue is resolved.
> 
> Update I have now solved the LLVM_DIR and Clang_DIR issue. Now I need to solve the following issue with the cmake configure stage of building CppInterOp
> 
> ```
> -- Looking for os_signpost_interval_begin
> CMake Error at D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2 (set):
>   Syntax error in cmake code at
> 
>     D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2
> 
>   when parsing string
> 
>     D:/a/CppInterOp/CppInterOp/cmake;D:/a/CppInterOp/CppInterOp/cmake/modules;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:\a\CppInterOp\CppInterOp\llvm-project\build\lib\cmake\llvm
> 
>   Invalid character escape '\a'.
> ```

Don't we need to use backslashes? Additionally, we can increase the 1 minute tmate session to more time [here](https://github.com/compiler-research/CppInterOp/blob/16f1c4bbf724e4308b2a0a618b61d8b2bbcea068/.github/workflows/ci.yml#L525) so that you can ssh into that machine and try things interactively.

**Comment by mcbarton** [comment]
> > > @alexander-penev @vgvassilev Currently I've got the Windows ci build upto the point where it needs to build CppInterOp , but it fails to find the llvm config file. I've tried a few different definitions of LLVM_DIR without success. Any ideas what the folder location should be on Windows?
> > > A separate issue which also needs to be fixed is that currently for the Windows ci build LLVM_HASH and CLING_HASH are currently hardcoded (I have an idea for a fix, but have yet to test). After these 2 issues have been fixed then this project should have a working ci which shows that CppInterOp builds on Windows, assuming no other issues arise once LLVM_DIR issue is resolved.
> > 
> > 
> > Update I have now solved the LLVM_DIR and Clang_DIR issue. Now I need to solve the following issue with the cmake configure stage of building CppInterOp
> > ```
> > -- Looking for os_signpost_interval_begin
> > CMake Error at D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2 (set):
> >   Syntax error in cmake code at
> > 
> >     D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2
> > 
> >   when parsing string
> > 
> >     D:/a/CppInterOp/CppInterOp/cmake;D:/a/CppInterOp/CppInterOp/cmake/modules;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:\a\CppInterOp\CppInterOp\llvm-project\build\lib\cmake\llvm
> > 
> >   Invalid character escape '\a'.
> > ```
> 
> Don't we need to use backslashes? Additionally, we can increase the 1 minute tmate session to more time [here](https://github.com/compiler-research/CppInterOp/blob/16f1c4bbf724e4308b2a0a618b61d8b2bbcea068/.github/workflows/ci.yml#L525) so that you can ssh into that machine and try things interactively.

Truth be told I don't know. I never build on Windows, so most of this PR has been guess work based on what I've been looking at online. My guess is that its not supposed to be \ given it has no issue with any of the other path directories.

**Comment by mcbarton** [comment]
> > > @alexander-penev @vgvassilev Currently I've got the Windows ci build upto the point where it needs to build CppInterOp , but it fails to find the llvm config file. I've tried a few different definitions of LLVM_DIR without success. Any ideas what the folder location should be on Windows?
> > > A separate issue which also needs to be fixed is that currently for the Windows ci build LLVM_HASH and CLING_HASH are currently hardcoded (I have an idea for a fix, but have yet to test). After these 2 issues have been fixed then this project should have a working ci which shows that CppInterOp builds on Windows, assuming no other issues arise once LLVM_DIR issue is resolved.
> > 
> > 
> > Update I have now solved the LLVM_DIR and Clang_DIR issue. Now I need to solve the following issue with the cmake configure stage of building CppInterOp
> > ```
> > -- Looking for os_signpost_interval_begin
> > CMake Error at D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2 (set):
> >   Syntax error in cmake code at
> > 
> >     D:/a/CppInterOp/CppInterOp/build/CMakeFiles/CMakeScratch/TryCompile-j9evml/CMakeLists.txt:2
> > 
> >   when parsing string
> > 
> >     D:/a/CppInterOp/CppInterOp/cmake;D:/a/CppInterOp/CppInterOp/cmake/modules;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:/a/CppInterOp/CppInterOp/llvm-project/build/lib/cmake/llvm;D:\a\CppInterOp\CppInterOp\llvm-project\build\lib\cmake\llvm
> > 
> >   Invalid character escape '\a'.
> > ```
> 
> Don't we need to use backslashes? Additionally, we can increase the 1 minute tmate session to more time [here](https://github.com/compiler-research/CppInterOp/blob/16f1c4bbf724e4308b2a0a618b61d8b2bbcea068/.github/workflows/ci.yml#L525) so that you can ssh into that machine and try things interactively.

Looking at the cmake configure stage for llvm on windows they have / not \  (example from ci shown below)
```
-- The C compiler identification is MSVC 19.37.32826.1
-- The CXX compiler identification is MSVC 19.37.32826.1
-- The ASM compiler identification is MSVC
-- Found assembler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.37.32822/bin/HostX64/x64/cl.exe
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.37.32822/bin/HostX64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.37.32822/bin/HostX64/x64/cl.exe - skipped
```

**Comment by mcbarton** [comment]
I did a little research and found out that although Windows paths have \ as far as cmake as is concerned the paths should always use / , as in Windows a \ is likely to interpreted as escape sequence. So to solve the current issue it needs to be determined in the cmake files the \ path is being included, and change it to use / (references below)
https://discourse.cmake.org/t/window-env-variable-and-cmake-path/5410
https://stackoverflow.com/questions/28070810/cmake-generate-error-on-windows-as-it-uses-as-escape-seq

**Comment by mcbarton** [comment]
@vgvassilev I have now got past the parsing issue for the clang-repl builds of CppInterOp on Windows (worked around it by replacing \ with / in LLVM_DIR within cmakelists.txt). The issue with them now seems to be a few header files it can't find, and a few other errors such as . I think some more familiar with the CppInterOp code and Clang needs to fix those issues.

The cling build still seems to be having issues with finding the cling config file. Not entirely sure what is causing that issue.

**Comment by vgvassilev** [comment]
> @vgvassilev I have now got past the parsing issue for the clang-repl builds of CppInterOp on Windows (worked around it by replacing \ with / in LLVM_DIR within cmakelists.txt). The issue with them now seems to be a few header files it can't find, and a few other errors . I think some more familiar with the CppInterOp code and Clang needs to fix those issues.
> 
> The cling build still seems to be having issues with finding the cling config file. Not entirely sure what is causing that issue.

We might need to include `llvm/Support/Casting.h` for the `isa` to work?

**Comment by mcbarton** [comment]
@vgvassilev can you delete the cling llvm cache from the windows build by hand? Looking at the ci when it was last built the cling directory was incorrect, so it was never built, so that why it can't be found.

**Comment by vgvassilev** [comment]
Done.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -660,6 +660,17 @@ jobs:
         echo "CPPINTEROP_DIR=$env:CPPINTEROP_DIR"
         echo "CPPINTEROP_DIR=$env:CPPINTEROP_DIR" >> $env:GITHUB_ENV  
 
+        #Build dlfcn-win32 so can use dlfcn.h on Windows
+        git clone --depth=1 https://github.com/dlfcn-win32/dlfcn-win32.git
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:664]: We should "just" use the relevant windows API. We do not need the dlfcn specifically.

Comment (mcbarton) [reply .github/workflows/ci.yml:664]: Ok. I will revert back to the old method. I only saw this comment after I put in the last commit.

**Comment by fsfod** [comment]
I don't know how much use this, but been experimenting with getting windows to build. I have it successfully building with few things commented out. For dlfcn\dlopen i just used some of llvm support API's. You see my changes here https://github.com/fsfod/CppInterOp/tree/windows

File: .github/workflows/ci.yml (hunk 3 above)
Comment (mcbarton) [reply .github/workflows/ci.yml:664]: All windows builds are currently at the same stage now where errors such as missing header etc are stopping the build.

My attempts locally to work out how to get the LLVM_HASH and Cling_HASH without hardcoded have been unsuccessful

Comment (mcbarton) [reply .github/workflows/ci.yml]: I have reverted these changes now.

**Comment by vgvassilev** [comment]
@fsfod, this looks quite nice. Are you interested in moving some of that work upstream?

**Comment by mcbarton** [comment]
> I don't know how much use this, but been experimenting with getting windows to build. I have it successfully building with few things commented out. For dlfcn\dlopen i just used some of llvm support API's. You see my changes here https://github.com/fsfod/CppInterOp/commits/windows/

Thanks. I would also appreciate if your work is moved upstream, like @vgvassilev asks.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (fsfod) [reply .github/workflows/ci.yml:284]: maybe %LLVM_HASH%

**Comment by vgvassilev** [comment]
What’s the plan here? Do we want to merge that PR and move to the other one with the windows changes in the codebase?

**Comment by mcbarton** [comment]
> What’s the plan here? Do we want to merge that PR and move to the other one with the windows changes in the codebase?

@vgvassilev Yes. That is the plan. I thought if this PR gets merged then the changes in the other PR can be checked. I've started the process of setting myself up a build environment on local Windows machine, so I can try and help with the other PR.

**Comment by mcbarton** [comment]
> > What’s the plan here? Do we want to merge that PR and move to the other one with the windows changes in the codebase?
> 
> @vgvassilev Yes. That is the plan. I thought if this PR gets merged then the changes in the other PR can be checked. I've started the process of setting myself up a build environment on a local Windows machine, so I can try and help with the other PR. This PR will help to determine when the other PR is ready for merging I believe.

Assuming your ok for this idea then this PR is ready for review.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -21,102 +21,191 @@ jobs:
...
+          
           - name: osx13-x86-clang-clang-repl-16
             os: macos-13
             compiler: clang
             clang-runtime: '16'
             cling: Off
-            cppyy: Off
-
+            cppyy: Off 
+
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:185]: Let's clean up the whitespace changes if they are irrelevant.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -21,102 +21,191 @@ jobs:
...
+            coverage: true
+            
+          - name: ubu22-x86-gcc9-clang-repl-16
             os: ubuntu-22.04
             compiler: gcc-9
             clang-runtime: '16'
             cling: Off
             cppyy: Off
             coverage: true
-            extra_cmake_options: '-DLLVM_USE_LINKER=lld'
```
[the remark this answers is not in the document]
Comment (mcbarton) [reply .github/workflows/ci.yml:47]: I dropped extra_cmake_options as it wasn't being used. I believe I can add LLVM_USE_LINKER  to all jobs (it works on my Macbook), but I think LLVM_ENABLE_LLD=ON is better, as from my experience you need it to be built when compiling llvm with gcc on osx, rather then clang which is used in this ci. Are you ok with me using LLVM_ENABLE_LLD?

[the remark this answers is not in the document]
Comment (mcbarton) [reply .github/workflows/ci.yml:47]: To test that LLVM_ENABLE_LLD works will all jobs I will need the cache cleared.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (mcbarton) [reply .github/workflows/ci.yml]: White space removed with latest commit.

File: .github/workflows/ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (vgvassilev) [reply .github/workflows/ci.yml:47]: Ok, done.

**Comment by mcbarton** [comment]
@alexander-penev This PR is ready for review. Applying the patches to clang is currently hardcoded for Windows, but I will fix this in a later PR. I can't explain the spaces in the patch file, except for without them the patch will fail to apply on Windows (it will output that it finds the patch file corrupt every time it meets a blank line without them). The windows ci build is not mean to pass. This PR is supposed to put in a place a ci build for Windows, with the codebase changes in https://github.com/compiler-research/CppInterOp/pull/181 meant to help it pass.

File: .github/workflows/ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcbarton) [reply .github/workflows/ci.yml:47]: This has been fixed with the new LLVM_ENABLE_LLD passing.

**Review by alexander-penev — APPROVED** [review]

**Pull request merged**